src/vanilla/store.ts

```diff
@@ -725,43 +710,88 @@ const buildStore = (
+  const derivedStore = store.unstable_derive(
+    (getAtomState, atomRead, atomWrite, atomOnMount) => {
+      savedGetAtomState = getAtomState
+      return [
+        (atom) => {
+          let proxyAtomState = proxyAtomStateMap.get(atom)
+          if (!proxyAtomState) {
+            const atomState = getAtomState(atom)
+            proxyAtomState = new Proxy(atomState, {
+              set(target, prop, value) {
+                if (prop === 'm') {
+                  debugMountedAtoms.add(atom)
+                }
+                return Reflect.set(target, prop, value)
+              },
+              deleteProperty(target, prop) {
+                if (prop === 'm') {
+                  debugMountedAtoms.delete(atom)
+                }
+                return Reflect.deleteProperty(target, prop)
+              },
+            })
+            proxyAtomStateMap.set(atom, proxyAtomState)
           }
-          return atomState
+          return proxyAtomState
         },
-      }),
-      dev4_get_mounted_atoms: () => debugMountedAtoms,
-      dev4_restore_atoms: (values) => {
-        const batch = createBatch()
-        for (const [atom, value] of values) {
-          if (hasInitialValue(atom)) {
-            const atomState = getAtomState(atom)
-            const prevEpochNumber = atomState.n
-            setAtomStateValueOrPromise(atom, atomState, value)
-            mountDependencies(batch, atom, atomState)
-            if (prevEpochNumber !== atomState.n) {
-              registerBatchAtom(batch, atom, atomState)
-              recomputeDependents(batch, atom, atomState)
+        atomRead,
+        (atom, getter, setter, ...args) => {
+          if (RESTORE_ATOM in atom) {
+            const values = atom[RESTORE_ATOM] as Iterable<
+              readonly [AnyAtom, AnyValue]
+            >
+            for (const [atom, value] of values) {
+              if (hasInitialValue(atom)) {
+                setter(atom as never, value)
+              }
             }
+            return undefined as never
           }
+          return atomWrite(atom, getter, setter, ...args)
+        },
+        atomOnMount,
+      ]
+    },
+  )
+  const devStore: DevStoreRev4 = {
+    // store dev methods (these are tentative and subject to change without notice)
+    dev4_get_internal_weak_map: () => ({
+      get: (atom) => {
+        const atomState = savedGetAtomState(atom)
+        if (atomState.n === 0) {
+          // for backward compatibility
+          return undefined
         }
-        flushBatch(batch)
+        return atomState
       },
-    }
-    Object.assign(store, devStore)
+    }),
+    dev4_get_mounted_atoms: () => debugMountedAtoms,
+    dev4_restore_atoms: (values) => {
+      const restoreAtom: WritableAtom<null, [], void> &
+        Record<typeof RESTORE_ATOM, typeof values> = {
+        [RESTORE_ATOM]: values,
+        read: () => null,
+        write: () => {},
+      }
+      derivedStore.set(restoreAtom)
```

Comment:
I'm curious why not simply?
```ts
store.set(atom(null, (get, set) => {
  values.forEach((args) => set(...args))
})
```

Reply:
`atom` in `values` can have a custom `write` function, which we need to bypass.